### PR TITLE
Fix and improve ViewDefs

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/core/src/view/provider.ts [hunk 1]
```diff
@@ -432,25 +432,41 @@ export function callLifecycleHooksChildrenFirst(view: ViewData, lifecycles: Node
...
+    let parent = nodeDef.parent;
+    if (!parent && nodeDef.flags & lifecycles) {
+      // matching root node (e.g. a pipe)
+      callProviderLifecycles(view, i, nodeDef.flags & lifecycles);
+    }
+    if ((nodeDef.childFlags & lifecycles) === 0) {
+      // no child matches one of the lifecycles
       i += nodeDef.childCount;
     }
+    while (parent && (parent.flags & NodeFlags.TypeElement) &&
```
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: check `lifecycles` inside.

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts [hunk 2]
```diff
@@ -800,6 +800,7 @@ function createTests({viewEngine}: {viewEngine: boolean}) {
       });
 
       describe('ngAfterContentInit', () => {
+        // TODO: it should call siblings in the right order...
```
Comment (tbosch) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: remove

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts [hunk 3]
```diff
@@ -1045,8 +1047,8 @@ function createTests({viewEngine}: {viewEngine: boolean}) {
              });
 
              const ctx = createCompFixture(
-                 '<div testDirective="parent"><div *ngFor="let x of [0,1]" testDirective="contentChild{{x}}"></div>' +
-                     '<other-cmp></other-cmp></div>',
+                 '<divp testDirective="parent"><div *ngFor="let x of [0,1]" testDirective="contentChild{{x}}"></div>' +
```
Comment (tbosch) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: remove `divp`

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts [hunk 4]
```diff
@@ -1061,13 +1063,13 @@ function createTests({viewEngine}: {viewEngine: boolean}) {
         it('should be called in reverse order so the child is always notified before the parent',
            fakeAsync(() => {
              const ctx = createCompFixture(
-                 '<div testDirective="parent"><div testDirective="child"></div></div>');
+                 '<divp testDirective="parent"><div testDirective="child"></div></divp><div testDirective="sibling"></div>');
```
Comment (tbosch) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: `divp`

**Review by vicb — APPROVED** [review]
LGTM when Travis is green

File: modules/@angular/core/src/view/provider.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: Done

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: Done

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: Done

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: Done

**Pull request merged**